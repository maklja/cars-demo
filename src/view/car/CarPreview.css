.car-preview {
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	border-radius: 1.5em;
	background-color: #eeeeee;
	height: 100%;
	width: 100%;
	min-height: 18.75em;
	max-height: 40.75em;
	overflow: hidden;
}

.car-preview.car-preview--selected {
	border: 0.6em solid #4caf50;
}

.car-preview .car-preview__card {

	/* Card container that contains front and back parts */
	height: 90%;
	width: 100%;
	position: relative;
	transform-style: preserve-3d;
	perspective: 500px; /* setup perspective for better 3D experience during fliping */
	transition: 0.6s;
	cursor: pointer;
}

.car-preview .car-preview__card .car-preview__front,
.car-preview .car-preview__card .car-preview__back {
	display: flex;
	align-items: center;
	justify-content: center;
	height: 100%;
	width: 100%;
	backface-visibility: hidden; /* dont make back of the cards visible after rotation */
	transition: 0.6s;
	transform-style: preserve-3d;
}

.car-preview .car-preview__card .car-preview__front {

	/* init state of the front part */
	transform: translateZ(0) rotateX(0deg);
	z-index: 2; /* make front card on top */

	/* IE support for transition with backface-visibility hidden is buggy and poorly implemented
       so we need to hack it with animations */
	-ms-animation: reverse-spin 0.6s; /* stylelint-disable-line property-no-vendor-prefix */
}

.car-preview .car-preview__card:hover .car-preview__front {

	/* transformation after hover on front part */
	transform: translateZ(-400px) rotateX(180deg); /* on hover start moving card to the back and rotate it around x axis */
	-ms-animation: spin 0.6s; /* stylelint-disable-line property-no-vendor-prefix */
}

.car-preview .car-preview__card .car-preview__back {

	/* init state of back part */
	transform: translateZ(-400px) rotateX(-180deg);
	position: absolute;
	top: 0;
	left: 0;
}

.car-preview .car-preview__card:hover .car-preview__back {

	/* transformation after hover on back part */
	transform: translateZ(0) rotateX(0deg); /* after hover rotate element to face the user */
	-ms-animation: reverse-spin 0.6s; /* stylelint-disable-line property-no-vendor-prefix */
}

.car-preview .car-preview__overlay {

	/* back of the card with dark transparent overlay */
	background: rgba(23, 23, 23, 0.5);
	position: absolute;
	top: 0;
	left: 0;
	width: 100%;
	height: 100%;
}

.car-preview .car-preview__description {

	/* car description on back part */
	position: absolute;
	top: 0;
	left: 0;
	width: 100%;
	height: 100%;
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	text-align: center;
	color: white;
	font-size: 2em;
	z-index: 1; /* position it on top of overlay */
	padding: 1.5em;
}

.car-preview .car-preview__image {
	max-width: 100%;
	flex-shrink: 0; /* fix for IE11 issue https://stackoverflow.com/questions/36822370/flexbox-on-ie11-image-stretched-for-no-reason */
}

.car-preview .car-preview__title {
	font-size: 3em;
}

/* IE workaround animations */
@keyframes spin {

	from {
		transform: translateZ(0) rotateX(0deg);
	}

	to {
		transform: translateZ(-400px) rotateX(180deg);
	}
}

@keyframes reverse-spin {

	from {
		transform: translateZ(-400px) rotateX(-180deg);
	}

	to {
		transform: translateZ(0) rotateX(0deg);
	}
}
