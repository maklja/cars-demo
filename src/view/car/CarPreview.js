import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { carPropTypes, carPropDefaults } from '../../props';

import './CarPreview.css';

class CarPreview extends Component {
	constructor(props) {
		super(props);

		this.state = {
			isSelected: props.isSelected
		};

		this._onClick = this._onClick.bind(this);
	}

	render() {
		const {
			id,
			speed,
			name,
			image,
			description,
			disableSelection
		} = this.props;
		const { isSelected } = this.state;

		return (
			<div
				className={`car-preview ${
					isSelected ? 'car-preview--selected' : '' // is selected put border around tile
				} ${disableSelection ? 'car-preview--disabled' : ''}`}
				data-id={id}
				onClick={this._onClick}
			>
				<div className="car-preview__card">
					<div className="car-preview__front">
						<img
							src={image}
							alt={name}
							className="car-preview__image"
						/>
					</div>
					<div className="car-preview__back">
						<div className="car-preview__overlay" />
						<div className="car-preview__description">
							<div className="car-preview__description-text">
								{description}
							</div>
							<div>
								Top speed: {speed}
								km/h
							</div>
						</div>
						<img
							src={image}
							alt={name}
							className="car-preview__image"
						/>
					</div>
				</div>
				<div className="car-preview__title">{name}</div>
			</div>
		);
	}

	_onClick() {
		const { id, onSelectionChanged, disableSelection } = this.props;

		// if selection is disable, just return
		if (disableSelection) {
			return;
		}

		this.setState(
			state => {
				return {
					isSelected: !state.isSelected
				};
			},
			() => onSelectionChanged(id, this.state.isSelected) // notify parent that selection on element has changed
		);
	}
}

CarPreview.propTypes = {
	...carPropTypes,
	isSelected: PropTypes.bool,
	onSelectionChanged: PropTypes.func,
	disableSelection: PropTypes.bool
};

CarPreview.defaultProps = {
	...carPropDefaults,
	isSelected: false,
	onSelectionChanged: () => {},
	disableSelection: false
};

export default CarPreview;
