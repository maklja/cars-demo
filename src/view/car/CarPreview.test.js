import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme'; // use enzyme to simulate user interaction
import Adapter from 'enzyme-adapter-react-16';

import CarPreview from './CarPreview';

// configure enzyme for react 16
configure({ adapter: new Adapter() });

it('CarPreview renders correctly with Mazda 3', () => {
	const car = {
		image:
			'https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png',
		speed: 235,
		description:
			'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
		name: 'Mazda 3',
		id: 2
	};

	const tree = renderer.create(<CarPreview {...car} />).toJSON();
	expect(tree).toMatchSnapshot();
});

it('CarPreview renders correctly with Mercedes Benz CLA', () => {
	const car = {
		image: 'http://www.nriol.com/images/benz-cla.png',
		speed: 280,
		description:
			'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
		name: 'Mercedes Benz CLA',
		id: 3
	};

	const tree = renderer.create(<CarPreview {...car} />).toJSON();
	expect(tree).toMatchSnapshot();
});

it('CarPreview renders correctly when selected', () => {
	const car = {
		image: 'http://www.nriol.com/images/benz-cla.png',
		speed: 280,
		description:
			'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
		name: 'Mercedes Benz CLA',
		id: 3
	};

	const tree = renderer
		.create(<CarPreview {...car} isSelected={true} />)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

it('test selection change callback on initially unselected component', done => {
	const car = {
		image: 'http://www.nriol.com/images/benz-cla.png',
		speed: 280,
		description:
			'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
		name: 'Mercedes Benz CLA',
		id: 3
	};

	// create shallow copy with enzyme
	const wrapper = shallow(
		<CarPreview
			{...car}
			onSelectionChanged={(id, isSelected) => {
				// callback is called every time selection is change
				expect(id).toBe(car.id); // first parameter must be car id
				expect(isSelected).toBeTruthy(); // second parameter must be selection flag
				done();
			}}
		/>
	);

	// simulate click on control
	wrapper.simulate('click');
});

it('test selection change callback on initially selected component', done => {
	const car = {
		image: 'http://www.nriol.com/images/benz-cla.png',
		speed: 280,
		description:
			'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
		name: 'Mercedes Benz CLA',
		id: 3
	};

	// create shallow copy with enzyme
	const wrapper = shallow(
		<CarPreview
			{...car}
			isSelected={true}
			onSelectionChanged={(id, isSelected) => {
				// callback is called every time selection is change
				expect(id).toBe(car.id); // first parameter must be car id
				expect(isSelected).toBeFalsy(); // second parameter must be selection flag
				done();
			}}
		/>
	);

	// simulate click on control
	wrapper.simulate('click');
});
