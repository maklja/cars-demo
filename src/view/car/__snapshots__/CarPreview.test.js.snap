// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarPreview renders correctly when selected 1`] = `
<div
  className="car-preview car-preview--selected "
  data-id={3}
  onClick={[Function]}
>
  <div
    className="car-preview__card"
  >
    <div
      className="car-preview__front"
    >
      <img
        alt="Mercedes Benz CLA"
        className="car-preview__image"
        src="http://www.nriol.com/images/benz-cla.png"
      />
    </div>
    <div
      className="car-preview__back"
    >
      <div
        className="car-preview__overlay"
      />
      <div
        className="car-preview__description"
      >
        <div
          className="car-preview__description-text"
        >
          The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.
        </div>
        <div>
          Top speed: 
          280
          km/h
        </div>
      </div>
      <img
        alt="Mercedes Benz CLA"
        className="car-preview__image"
        src="http://www.nriol.com/images/benz-cla.png"
      />
    </div>
  </div>
  <div
    className="car-preview__title"
  >
    Mercedes Benz CLA
  </div>
</div>
`;

exports[`CarPreview renders correctly with Mazda 3 1`] = `
<div
  className="car-preview  "
  data-id={2}
  onClick={[Function]}
>
  <div
    className="car-preview__card"
  >
    <div
      className="car-preview__front"
    >
      <img
        alt="Mazda 3"
        className="car-preview__image"
        src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
      />
    </div>
    <div
      className="car-preview__back"
    >
      <div
        className="car-preview__overlay"
      />
      <div
        className="car-preview__description"
      >
        <div
          className="car-preview__description-text"
        >
          The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.
        </div>
        <div>
          Top speed: 
          235
          km/h
        </div>
      </div>
      <img
        alt="Mazda 3"
        className="car-preview__image"
        src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
      />
    </div>
  </div>
  <div
    className="car-preview__title"
  >
    Mazda 3
  </div>
</div>
`;

exports[`CarPreview renders correctly with Mercedes Benz CLA 1`] = `
<div
  className="car-preview  "
  data-id={3}
  onClick={[Function]}
>
  <div
    className="car-preview__card"
  >
    <div
      className="car-preview__front"
    >
      <img
        alt="Mercedes Benz CLA"
        className="car-preview__image"
        src="http://www.nriol.com/images/benz-cla.png"
      />
    </div>
    <div
      className="car-preview__back"
    >
      <div
        className="car-preview__overlay"
      />
      <div
        className="car-preview__description"
      >
        <div
          className="car-preview__description-text"
        >
          The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.
        </div>
        <div>
          Top speed: 
          280
          km/h
        </div>
      </div>
      <img
        alt="Mercedes Benz CLA"
        className="car-preview__image"
        src="http://www.nriol.com/images/benz-cla.png"
      />
    </div>
  </div>
  <div
    className="car-preview__title"
  >
    Mercedes Benz CLA
  </div>
</div>
`;
